src/engine/gnc-pricedb.c

```diff
@@ -1345,68 +1356,147 @@ pricedb_remove_foreach_pricelist (gpointer key,
 
     ENTER("key %p, value %p, data %p", key, val, user_data);
 
-    /* The most recent price is the first in the list */
-    if (!data->delete_last)
-        node = g_list_next(node);
-
     /* now check each item in the list */
     g_list_foreach(node, (GFunc)check_one_price_date, data);
 
     LEAVE(" ");
 }
 
 static void
-pricedb_remove_foreach_currencies_hash (gpointer key,
-                                        gpointer val,
-                                        gpointer user_data)
+free_data (gpointer data)
 {
-    GHashTable *currencies_hash = (GHashTable *) val;
-
-    ENTER("key %p, value %p, data %p", key, val, user_data);
-    g_hash_table_foreach(currencies_hash,
-                         pricedb_remove_foreach_pricelist, user_data);
-    LEAVE(" ");
+    g_free (data);
 }
 
+static void
+gnc_pricedb_get_remove_dates (GHashTable* hash_dates, GDate period_begin, GDate period_end, PriceRemoveOptions leave)
+{
+    Recurrence *r;
+    GDate recurrance_date_old, recurrance_date_next;
+
+    GDateWeekday selected_day_of_week = G_DATE_FRIDAY;
+    GDate day_of_week_date;
+
+    day_of_week_date = period_begin;
+
+    r = g_new (Recurrence, 1);
+
+    DEBUG("Period Begin date is %d/%d/%d and End date is %d/%d/%d", g_date_get_day (&period_begin),
+           g_date_get_month (&period_begin), g_date_get_year (&period_begin), g_date_get_day (&period_end),
+           g_date_get_month (&period_end), g_date_get_year (&period_end));
+
+    g_date_set_day (&day_of_week_date, 1); // set date to 1st of month
+
+    while (g_date_get_weekday (&day_of_week_date) != selected_day_of_week)
+        g_date_subtract_days (&day_of_week_date, 1); // go back till we find last Friday of month
```

Comment:
This is confusing me: your comment says you will go back till the last Friday of month. However this is at best the last Friday of the **previous** month or it will be the first day of **this** month in case it happens to be a Friday.
Is this what you intend ?